python/ray/data/dataset.py

```diff
@@ -902,6 +904,17 @@ def transform(block: Block, fn: RowUDF[T, U]) -> Iterable[Block]:
         plan = self._plan.with_stage(
             OneToOneStage("flat_map", transform, compute, ray_remote_args, fn=fn)
         )
+
+        logical_plan = self._logical_plan
+        if logical_plan is not None:
+            op = FlatMap(
+                input_op=logical_plan.dag,
+                block_fn=transform,
+                fn=fn,
+                compute=compute,
+                ray_remote_args=ray_remote_args,
+            )
+            logical_plan = LogicalPlan(op)
         return Dataset(plan, self._epoch, self._lazy)
```

Comment:
return Dataset(plan, self._epoch, self._lazy, logical_plan)

Reply:
To capture this bug, let's also add test for `flatmap()` and `filter()` in `test_execution_optimizer.py:test_e2e_optimizer_sanity()`. Thanks.

Reply:
good catch, updated and added test